core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -186,8 +186,23 @@ private[spark] class TaskSetManager(
 
   // Add all our tasks to the pending lists. We do this in reverse order
   // of task index so that tasks with low indices get launched first.
-  for (i <- (0 until numTasks).reverse) {
-    addPendingTask(i)
+  addPendingTasks()
+
+  private def addPendingTasks(): Unit = {
+    val (_, duration) = Utils.timeTakenMs {
+      for (i <- (0 until numTasks).reverse) {
+        addPendingTask(i, resolveRacks = false)
+      }
+      // Resolve the rack for each host. This can be slow, so de-dupe the list of hosts,
+      // and assign the rack to all relevant task indices.
+      val racks = sched.getRacksForHosts(pendingTasksForHost.keySet.toSeq)
```

Comment:
There's an implicit assumption here that `map.keySet` and `map.values` iterate in the same order. I'm not sure if that's guaranteed, and at the same time I don't see why that wouldn't be the case, but just wanted to point this out.

Reply:
I had the exact same thought when I reached that line. 
Even thought about a possible solutions: 

- creating a new val with the value `racks.entrySet` and generating the keys and values from this entry set (as in entry set the key and value is bound together the ordering will be fixed; even with one iteration the key and the value can be generated). 
- Another possible and more elegant solution is calling `racks.asScala.unzip`. 

Both solutions has some performance cost.

Reply:
Correction for last sentence: the first solution can be done without performance cost but probably the code will be a bit less elegant.

Reply:
great point, I have updated this to use `unzip`